Packet++/src/SomeIpLayer.cpp

```diff
@@ -0,0 +1,379 @@
+#include "PayloadLayer.h"
+#include <algorithm>
+#include <sstream>
+#include <string.h>
+
+namespace pcpp
+{
+std::unordered_set<uint16_t> SomeIpLayer::mSomeIpPorts{};
+
+void SomeIpLayer::optionallyCopyPayloadData(const uint8_t *const data, size_t dataLen, bool bDontCopy)
+{
+	if (bDontCopy)
+	{
+		m_DataLen = SOMEIP_HDR_LEN;
+	}
+	else
+	{
+		m_DataLen = SOMEIP_HDR_LEN + dataLen;
+	}
+
+	m_Data = new uint8_t[m_DataLen];
+	m_Protocol = SomeIp;
+	memset(m_Data, 0, SOMEIP_HDR_LEN);
+
+	if (bDontCopy)
+	{
+		m_PayloadPtr = (uint8_t *)data;
+	}
+	else
+	{
+		memcpy(m_Data + SOMEIP_HDR_LEN, data, dataLen);
+		m_PayloadPtr = nullptr;
+	}
+
+	setPayloadLength(dataLen);
+}
+
+SomeIpLayer::SomeIpLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID,
+						 uint8_t interfaceVersion, MsgType type, uint8_t returnCode, const uint8_t *const data,
+						 size_t dataLen, bool bDontCopy)
+{
+	optionallyCopyPayloadData(data, dataLen, bDontCopy);
+	setServiceID(serviceID);
+	setMethodID(methodID);
+	setClientID(clientID);
+	setSessionID(sessionID);
+	setInterfaceVersion(interfaceVersion);
+	setMessageType(type);
+	setReturnCode(returnCode);
+}
+
+SomeIpLayer::SomeIpLayer(const SomeIpLayer &other) : Layer(other)
+{
+	copyLayerData(other);
+}
+
+void SomeIpLayer::copyLayerData(const SomeIpLayer &other)
+{
+	m_PayloadPtr = other.m_PayloadPtr;
+}
+
+SomeIpLayer &SomeIpLayer::operator=(const SomeIpLayer &other)
+{
+	Layer::operator=(other);
+
+	copyLayerData(other);
+
+	return *this;
+}
+
+void SomeIpLayer::setPayloadLength(uint32_t payloadLength)
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+	hdr->length = htobe32(sizeof(someiphdr) - sizeof(hdr->serviceID) - sizeof(hdr->methodID) - sizeof(hdr->length) +
+						  payloadLength);
+}
+
+uint32_t SomeIpLayer::getHeaderLengthWithPayload() const
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	/* Add outer header of someip to length (messageId and length field itself) */
+	return sizeof(uint32_t) * 2 + be32toh(hdr->length);
+}
+
+size_t SomeIpLayer::getHeaderLen() const
+{
+	if (subProtocolFollows())
+	{
+		/* If there is a protocol coming after the normal SomeIp header, return just the header length so PcapPlusPlus
+	handles the oncoming protocols correctly */
+		return sizeof(someiphdr);
+	}
+	else
+	{
+		/* Else, return the length of the header + payload, since it is a normal SomeIp PDU */
+		return getHeaderLengthWithPayload();
+	}
+}
+
+bool SomeIpLayer::subProtocolFollows() const
+{
+	// TODO: handle SomeIPSD/SomeIPTP packets
+
+	return false;
+}
+
+void SomeIpLayer::computeCalculateFields()
+{
+	someiphdr *hdr = (someiphdr *)m_Data;
+
+	hdr->protocolVersion = 0x01;
+
+	if (!m_NextLayer)
+		return;
+
+	// TODO: handle SomeIPSD/SomeIPTP packets
+}
+
+bool SomeIpLayer::isSomeIpPort(uint16_t port)
+{
+	return std::any_of(mSomeIpPorts.begin(), mSomeIpPorts.end(),
+					   [&](const uint16_t &someIpPort) { return someIpPort == port; });
+}
+
+void SomeIpLayer::addSomeIpPort(uint16_t port)

```

Comment:
Just trying to understand. Is it expected that one who uses the implementation should first set some range of ports used for SOME/IP communication?

Reply:
Hi Michael, thanks for having time to take a look at the code.
Our current implementation works like this: We get a SOME/IP-SD packet and read the endpoints from it. We add the ports of the endpoints to this list of ports, so further SOME/IP communication on these ports is detected correctly. We could also do this autonomously in the SOME/IP-SD layer, but our first implementation outsourced this action to the application. 
Since SOME/IP-SD traffic is usually sent on the port 30490, we are checking for that port specifically, so we do not miss the first SOME/IP-SD packets.

What do you think about it?

Reply:
Hi @maruu,

Sorry for the late reply. I was thinking as well about how to detect SOME/IP traffic properly. You have an interesting way of doing it. I do not have anything against this approach at the moment. ;)